libraries/Microsoft.Bot.Builder.Azure/AzureBlobStorage.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.IO;
+using System.Linq;
+using System.Net;
+using System.Threading.Tasks;
+using System.Web;
+using Microsoft.Bot.Builder.Core.Extensions;
+using Microsoft.WindowsAzure.Storage;
+using Microsoft.WindowsAzure.Storage.Blob;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.Azure
+{
+    /// <summary>
+    /// Models IStorage using Azure Storge Blobs
+    /// </summary>
+    /// <remarks>
+    /// The AzureBlobStorage implements State's IStorage using a single Azure Storage Blob Container.
+    /// Each entity or StoreItem is serialized into a JSON string and stored in an individual text blob.
+    /// Each blob is named after the StoreItem key which is encoded and ensure it conforms a valid blob name.
+    /// Concurrency is managed in a per entity (e.g. per blob) basis. If an entity implement IStoreItem
+    /// its eTag property value will be set with the blob's ETag upon Read. Afterward an AccessCondition
+    /// with the ETag value will be generated during Write. New entities will simple have an null ETag.
+    /// </remarks>
+    public class AzureBlobStorage : IStorage
+    {
+        private readonly static JsonSerializerSettings SerializationSettings = new JsonSerializerSettings()
+        {
+            // we use all so that we get typed roundtrip out of storage, but we don't use validation because we don't know what types are valid
+            TypeNameHandling = TypeNameHandling.All
+        };
+
+        private readonly static JsonSerializer JsonSerializer = JsonSerializer.Create(SerializationSettings);
+
+        /// <summary>
+        /// The Azure Storage Blob Container where entities will be stored
+        /// </summary>
+        private Lazy<ValueTask<CloudBlobContainer>> Container { get; set; }

```

Comment:
I don't agree with Drew on this one - ValueTask<> vs Task<>. 

Not because he's right or wrong... as he's probalby right. 

But simply because I had never heard of ValueTask, and had to go look it up. Any other developers coming in here will probably be in the same situation. Simplicty seems more important than correctness in this case. 

Thoughts?